### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by nhamlv-55**
add debug printing
add api to mark node as incomplete or global
mark formal arguemtn in local pass as incomplete

### CODE REVIEW

File: src/Graph.cc [hunk 1]
```diff
@@ -938,7 +940,7 @@ bool sea_dsa::Graph::computeCalleeCallerMapping(
       Cell &nc = callerG.mkCell(*kv.first, Cell());
       if (!simMap.insert(c, nc)) {
         if (reportIfSanityCheckFailed) {
-          errs() << "ERROR 1: callee is not simulated by caller at "
```
Comment (kuhar) [inline src/Graph.cc]: seems out of sync with the main branch -- you can fix that by rebasing to stay on top of it

Comment (agurfinkel) [reply src/Graph.cc]: @nhamlv-55 : also run `clang-format` and split into several commits based on what is being added in the commit

Comment (agurfinkel) [reply src/Graph.cc]: https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

File: src/DsaLocal.cc [hunk 2]
```diff
@@ -1030,6 +1040,7 @@ bool Local::runOnModule(Module &M) {
   m_dl = &M.getDataLayout();
   m_tli = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI();
   m_allocInfo = &getAnalysis<AllocWrapInfo>();
+  errs() << "DSA Module: " << M.getName() << "\n";
```
Comment (agurfinkel) [inline src/DsaLocal.cc]: protect by `LOG`

File: src/DsaLocal.cc [hunk 3]
```diff
@@ -1040,7 +1051,7 @@ bool Local::runOnFunction(Function &F) {
   if (F.isDeclaration() || F.empty())
     return false;
 
-  LOG("progress", errs() << "DSA: " << F.getName() << "\n";);
+  errs() << "DSA: " << F.getName() << "\n";
```
Comment (agurfinkel) [inline src/DsaLocal.cc]: protect by `LOG`

File: include/sea_dsa/Graph.hh [hunk 4]
```diff
@@ -520,6 +520,14 @@ public:
     m_nodeType.heap = v;
     return *this;
   }
+  Node &setIncomplete(bool v = true) {
```
Comment (agurfinkel) [inline include/sea_dsa/Graph.hh]: in addition to `setIncomplete` you need `markIncomplete`. If a node is incomplete, then any node reachable from it is incomplete as well. `setIncomplete` just sets the flag. `markIncomplete` traverses the graph and propagates the marking

File: src/DsaLocal.cc [hunk 5]
```diff
@@ -974,6 +980,9 @@ void LocalAnalysis::runOnFunction(Function &F, Graph &g) {
         n.addAllocSite(a);
       // -- mark node as a stack node
       n.setAlloca();
+      // -- mark node as incomplete (formal arguments)
+      n.setIncomplete();
```
Comment (agurfinkel) [inline src/DsaLocal.cc]: this should be `markIncomplete` to spread the marking. The call should be done after `LocalAnalysis` terminates to ensure that everything reachable by an argument at the END is marked incomplete.

**Comment by nhamlv-55** [comment]
The newest commit still just setIncomplete instead of markIncomplete. I will change it asap

**Pull request closed without merging**